src/payment/data/sagas.test.js

```diff
@@ -863,6 +872,225 @@ describe('saga tests', () => {
+      }));
+
+    const testPlan = [...successfulTestPlan, ...runtimeFailureTestPlan];
+
+    for (let i = 0, test = testPlan[i]; testPlan.length > i; i++, test = testPlan[i]) {
+      const nt = test.validToPoll ? '' : 'n\'t';
+      const suffix = test.shouldFail ? ', but should fail becase of a missing value' : ', and succeed';
+      it(`PAYMENT_STATE.${test.stateKey} should${nt} poll${suffix}`, async () => {
+        const localDispatched = [];
+        const localCaughtErrors = [];
+        axiosMock.reset();
+
+        axiosMock.onGet(CC_PAYMENT_STATE_ENDPOINT).reply(
+          test.shouldFail ? 404 : 200,
+          { state: PAYMENT_STATE.COMPLETED },
+        );
+
+        const localSagaOptions = {
+          dispatch: action => localDispatched.push(action),
+          onError: err => localCaughtErrors.push(err),
+        };
+
+        try {
+          await runSaga(
+            {
+              getState: () => ({
+                payment:
+                    {
+                      basket: test.inputState,
+                    },
+              }),
+              ...localSagaOptions,
+            },
+            handlePaymentState,
+          ).toPromise();
+        } catch (e) {} // eslint-disable-line no-empty
+
+        expect(localDispatched).toEqual(test.expectedResult);
+        expect(localCaughtErrors).toEqual([]);
+      });
+    }
+
+    it(`Should Retry HTTP Errors ${DEFAULT_PAYMENT_STATE_POLLING_MAX_ERRORS}x then fail`, async () => {
+      mergeConfig({
+        PAYMENT_STATE_POLLING_DELAY_SECS: 0.1,
+      });
+
+      const localDispatched = [];
+      const localCaughtErrors = [];
+      axiosMock.reset();
+
+      axiosMock.onGet(CC_PAYMENT_STATE_ENDPOINT).reply(404, null);
+
+      const editableState = ({
+        payment:
+            {
+              basket: generateTestBasketState(PAYMENT_STATE.PENDING, [{ paymentNumber: 'QAZWSX' }]),
+            },
+      });
+
+      const localSagaOptions = {
+        dispatch: action => {
+          localDispatched.push(action);
+          editableState.payment.basket = paymentState(editableState.payment.basket, action);
+        },
+        onError: err => localCaughtErrors.push(err),
+      };
+
+      try {
+        await runSaga(
+          {
+            getState: () => editableState,
+            ...localSagaOptions,
+          },
+          handlePaymentState,
+        ).toPromise();
+      } catch (e) {} // eslint-disable-line no-empty
+
+      expect(localDispatched).toEqual([
+        ...Array.from({
+          length: DEFAULT_PAYMENT_STATE_POLLING_MAX_ERRORS,
+        }, () => (pollPaymentState.received({ state: PAYMENT_STATE.HTTP_ERROR }))),
+        ...expectRuntimeFailure(),
+      ]);
+      expect(localCaughtErrors).toEqual([]);
+
+      mergeConfig({
+        PAYMENT_STATE_POLLING_DELAY_SECS: DEFAULT_PAYMENT_STATE_POLLING_DELAY_SECS,
+      });
+    });
+
+    it('Should Retry until state changes', async () => {
+      mergeConfig({
```

Comment:
This test proves we will poll as long as we have to to move from Pending to a non polling state.